app/models/minion.rb

```diff
@@ -4,18 +4,43 @@
 
 # Minion represents the minions that have been registered in this application.
 class Minion < ApplicationRecord
+  class NotEnoughMinions < StandardError; end
+  class CouldNotAssignRole < StandardError; end
```

Comment:
How are these exceptions handled by controllers ? IIRC, they will be handled as an 500, no ? If that's the case, could you change the application_controller so these exceptions are simple 404 responses (they are not really a server failure).